### PR TITLE
Track recognizable objects used by VOM so other objects can be GC'd

### LINKED ISSUE #3161: VOM should track recognizable vrefs (conservatively), inhibit syscall.retireImport
## What is the Problem Being Solved?

#3106 is about reducing RAM usage for unreachable objects. The next phase after that is to reduce disk usage by pruning unnecessary DB entries, starting with c-list entries for objects that no longer need their identity tracked.

The low-hanging fruit is the vref for a Presence which was imported into a vat, never used as the key of a WeakMap or WeakSet, and is then released. This will trigger a `syscall.dropImport` when the Presence is collected, but it could also trigger `syscall.retireImport`. That would allow the kernel to delete the importing vat's c-list entry and reduce the recognizability count. Once that count drops to zero, the kernel could notify the exporting vat with `dispatch.retireExport` and then delete the export-side c-list entry.

To do this reliably, the virtual object manager must be able to tell liveslots whether a given unreachable vref is recognizeable or not. We'll start with a conservative approach that simply remembers (in a strong Set) whether a given vref has *ever* been used as the key of a WeakMap/WeakSet. If so, the vat never retires the import. This conservative approach would never remove the item, even if the entry was manually deleted from every WeakMap/WeakSet, and/or all those maps/sets were collected themselves.

This should let us retire the vast majority of vrefs (we don't use *any* Presences as weak-collection keys right now, only Representatives, in the ERTP `paymentLedger`). However it would create a RAM burden of one vref string per `Payment` object, because the `Set` lives in RAM.

The second step is to move the flag onto disk, removing the RAM burden and replacing it with a (somewhat larger) disk burden. Something like `vatstoreSet('${vref}.recognizable', 1)`. I think we're using more Presences overall than Payments, so expect it would be a net improvement, but we should have a way to measure it.

A much later stage would be to replace the conservative "once added, never removed" approach with more detailed refcounting. This might take the form of a list of weak-thing tables that use the key: `vatstoreSet('${vref}.recognizable', JSON.stringify(['weakmap1', 'weakset3']))`, which would be most efficient if userspace tends to have a small number of weak tables (and it's ok if they hold a very large number of keys). Or we flip it around, use one vatstore key per weak table, and give each one a list of vrefs used as keys (better if userspace has a large number of weak tables, but bad if they have a large number of keys). Or we find some more efficient scheme. Maybe a Bloom Filter?

We won't need to implement any specific cycle-handling code to make this less conservative, however it won't retire as much as it possibly can until we do handle cycles in the virtualized-data reference graph, because otherwise the entries can't be deleted and their keys will remain recognizable.

## Test Plan

Normal unit tests: send a Presence into a vat, use it (or not) as a WeakMap and WeakSet key (test both!), tell the vat to drop its strong reference to the Presence, then watch to see whether the vat does both `syscall.dropImport` and `syscall.retireImport`, or only just `syscall.dropImport`.

### PR DESCRIPTION
**Pull request by FUDCo**
This was surprisingly straightforward and made the GC noticeably more aggressive, at least in the tests that it affected.

I think it needs more tests targeted at this feature specifically, but I'm not sure how to construct them, so I'm creating the PR now to see if reviewing suggests some.

Closes #3161

### CODE REVIEW

File: packages/SwingSet/src/kernel/virtualObjectManager.js
Comment (warner) [inline packages/SwingSet/src/kernel/virtualObjectManager.js]: Ah, right, because in the places that call this, they have a `vkey` but that's very much not the same thing as a `vref`.

Maybe we should refactor `virtualObjectKey` to take a `vref` instead? So the callers would do something like:
```js
const vref = getSlotForVal(key);
const vkey = virtualObjectKey(vref);
if (vkey) {
  addRecognizablePresence(vref);
  otherstuff(vkey);
  ..
```
?

File: packages/SwingSet/test/test-gc-kernel.js
Comment (warner) [inline packages/SwingSet/test/test-gc-kernel.js]: "The device should hold a reachable+recognizable reference, and vat-left (which forgot about amy) does not contribute either form of refcount, making the expected count 1,1. If deviceKeeper..."

**Review by warner — APPROVED** [review]
Minor changes recommended. Looks great, and will have a huge impact. Nice!

Please also add a "closes #3161" to the commit comment.

**Comment by warner** [comment]
Oh, actually the behavior may change when we land the #3406 branch (more drops happening), so please hold off on landing this until that change has landed and we fix any tests that break when these two great changes get together for a retirement party.

**Comment by warner** [comment]
> Oh, actually the behavior may change when we land the #3406 branch (more drops happening), so please hold off on landing this until that change has landed and we fix any tests that break when these two great changes get together for a retirement party.

That commit (8eeec2808ee7596d0b08a362d182c65a8828fba3) has landed, in PR #3416. Go ahead and test.

File: packages/SwingSet/src/kernel/virtualObjectManager.js
Comment (FUDCo) [reply packages/SwingSet/src/kernel/virtualObjectManager.js]: The whole point of `virtualObjectKey` is converting between an object reference and a key string with a vref as the intermediate state on the way there.  If it were to take a vref instead you'd just be adding some logic that would have to be replicated in a bunch of places.  The way it's factored now is right, I think, aside from the misleading parallel between the naming of `addReachablePresence` and `addRecognizablePresence`, so I think the proper approach is to address the naming directly.  I'm thinking `addReachablePresenceRef` and `addRecognizablePresenceValue`.

**Comment by FUDCo** [comment]
Changed one of the controller tests to make it less sensitive to GC variations, at the expense of checking less precisely.  Please re-review to decide if this is acceptable.

**Review by warner — APPROVED** [review]
Looks good. It's a shame to have to remove those checks, but they *were* kind of overspecified in the face of GC engine differences.

**Comment by warner** [comment]
Do please modify the commit comments to include a reference to the bug number, so we can track it down from the git history later.

**Pull request merged**
